### PR TITLE
Add a boolean expression parser

### PR DESCRIPTION
**Pull request by pchote**
This adds a quick and dirty (but still relatively efficient) parser for strings that define boolean expressions over runtime-defined variables.  It supports AND/OR/XOR/NOT and parentheses.  This will be used in future PRs that overhaul and simplify our prerequisite and upgrade declarations.

This comes with unit tests that verify correctness, so I'm hoping this will be a quick an easy review.

### CODE REVIEW

File: OpenRA.Game/Support/BooleanExpression.cs [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+			if (!tokens.Any())
+				throw new InvalidDataException("Empty expression");
+
+			if (closeParens != openParens)
+				throw new InvalidDataException("Mismatched opening and closing parenthesis");
+
+			for (var i = 0; i < tokens.Count - 1; i++)
+			{
+				if (tokens[i] is NotToken && !(tokens[i + 1] is VariableToken || tokens[i + 1] is OpenParenToken))
+					throw new InvalidDataException("Unexpected character `!` at index {0}".F(tokens[i].Index));
```
Comment (Phrohdoh) [inline OpenRA.Game/Support/BooleanExpression.cs]: Is there a concrete list of expected token types at the point of these exceptions?
If so those would be nice to include.

**Comment by pchote** [comment]
Rewritten to use the postfix formulation as requested by @chrisforbes in IRC.  I've kept the original recursive version at [pchote/original-boolexpr](https://github.com/pchote/OpenRA/tree/original-boolexpr) for comparison.

File: OpenRA.Test/OpenRA.Game/BinaryExpressionTest.cs [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+			AssertTrue("!false");
+		}
+
+		[TestCase(TestName = "Precidence")]
+		public void TestPrecidence()
+		{
+			AssertTrue("true && false || true");
+			AssertFalse("false || false && true");
+			AssertTrue("true && !true || !false");
+			AssertFalse("false || !true && !false");;
```
Comment (chrisforbes) [inline OpenRA.Test/OpenRA.Game/BinaryExpressionTest.cs]: Stray semicolon here.

File: OpenRA.Test/OpenRA.Game/BinaryExpressionTest.cs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+		}
+
+		[TestCase(TestName = "NOT operation")]
+		public void TestNOT()
+		{
+			AssertFalse("!true");
+			AssertTrue("!false");
+		}
+
+		[TestCase(TestName = "Precidence")]
```
Comment (chrisforbes) [inline OpenRA.Test/OpenRA.Game/BinaryExpressionTest.cs]: Precedence (everywhere)

**Review by chrisforbes — APPROVED** [review]

**Comment by pchote** [comment]
Fixed.

**Review by chrisforbes — APPROVED** [review]

**Review by phrohdoh — CHANGES_REQUESTED** [review]
Please fix `!!false`.

**Comment by pchote** [comment]
* Fixed `!!false` (adding plumbing for left vs right associativity) and added unit tests
* Added `==` operator and unit tests
* Renamed `^` to `!=`

**Review by phrohdoh — APPROVED** [review]
LGTM

**Pull request merged**